### PR TITLE
[tune] reuse_actors per default for function trainables

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Function trainables don't carry state, so they should be reused per default for performance optimization.

## Related issue number

Would e.g. simplify issues for frequently pausing trials like in #23470

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/tests/test_actor_reuse.py [hunk 1]
```diff
@@ -61,31 +58,6 @@ def reset_config(self, new_config):
...
-    trainable = wrap_function(trainable)
-
-    class ResetCountTrainable(trainable):
-        def reset_config(self, new_config):
-            num_resets[self.trial_id] += 1
-            return super().reset_config(new_config)
-
-    return ResetCountTrainable
-
-
```
Comment (krfricke) [inline python/ray/tune/tests/test_actor_reuse.py:88]: This test never worked, as `num_resets` was local to each remote actor.

There is no great way to count the number of resets for function trainables so I've removed this for now

File: python/ray/tune/tests/test_actor_reuse.py [hunk 2]
```diff
@@ -134,15 +101,6 @@ def testTrialReuseEnabled(self):
...
         self.assertEqual([t.last_result["num_resets"] for t in trials], [4, 5, 6, 7])
 
-    def testTrialReuseEnabledFunction(self):
-        num_resets = defaultdict(lambda: 0)
-        trials = self._run_trials_with_frequent_pauses(
-            create_resettable_function(num_resets), reuse=True
-        )
-        self.assertEqual([t.last_result["id"] for t in trials], [0, 1, 2, 3])
-        self.assertEqual([t.last_result["iter"] for t in trials], [2, 2, 2, 2])
-        self.assertEqual([num_resets[t.trial_id] for t in trials], [0, 0, 0, 0])
```
Comment (krfricke) [inline python/ray/tune/tests/test_actor_reuse.py:144]: As you can see, this test doesn't work - the number of resets should have been 4, 5, 6, 7

**Comment by xwjiang2010** [comment]
Is there concern if these function trainables use GPU?

File: python/ray/tune/tune.py [hunk 3]
```diff
@@ -455,6 +458,17 @@ def _handle_execute_queue():
...
         result_buffer_length = 1
 
+    # If reuse_actors is unset, default to False for string and class trainables,
+    # and default to True for everything else (i.e. function trainables)
+    if reuse_actors is None:
+        reuse_actors = not (
+            isinstance(run_or_experiment, str)
+            or (
+                inspect.isclass(run_or_experiment)
+                and issubclass(run_or_experiment, Trainable)
```
Comment (Yard1) [inline python/ray/tune/tune.py:468]: Can we actually make this a util function? `is_function_trainable` or something.

**Review by Yard1 — APPROVED** [review]
Looks good to me. CI seems to be failing though, unsure if related or not

**Comment by krfricke** [comment]
@xwjiang2010 thanks, that is indeed a concern, so lets default to False here

**Review by Yard1 — APPROVED** [review]

**Review by xwjiang2010 — APPROVED** [review]

**Pull request merged**